src/stdpar/plugin-SiPixelClusterizer/gpuClustering.h

```diff
@@ -1,292 +1,261 @@
+            auto p = hist.begin() + j;
+            auto i = *p + firstPixel;
+            auto m = clusterId[i];
+            while (m != clusterId[m])
+              m = clusterId[m];
+            clusterId[i] = m;
+          }
+        } else {
+          more = false;
+          for (uint32_t j = 0, k = 0U; j < hist.size(); ++j, ++k) {
+            auto p = hist.begin() + j;
+            auto i = *p + firstPixel;
+            for (int kk = 0; kk < nnn[k]; ++kk) {
+              auto l = nn[k][kk];
+              auto m = l + firstPixel;
+              assert(m != i);
+              auto old = cms::cuda::atomicMin(&clusterId[m], clusterId[i]);
+              if (old != clusterId[i]) {
+                // end the loop only if no changes were applied
+                more = true;
+              }
+              cms::cuda::atomicMin(&clusterId[i], old);
+            }  // nnloop
+          }    // pixel loop
         }
-      } else {
-        more = false;
-        for (auto j = threadIdx.x, k = 0U; j < hist.size(); j += blockDim.x, ++k) {
-          auto p = hist.begin() + j;
-          auto i = *p + firstPixel;
-          for (int kk = 0; kk < nnn[k]; ++kk) {
-            auto l = nn[k][kk];
-            auto m = l + firstPixel;
-            assert(m != i);
-            auto old = atomicMin(&clusterId[m], clusterId[i]);
-            if (old != clusterId[i]) {
-              // end the loop only if no changes were applied
-              more = true;
-            }
-            atomicMin(&clusterId[i], old);
-          }  // nnloop
-        }    // pixel loop
-      }
-      ++nloops;
-    }  // end while
+        ++nloops;
+      }  // end while
 
 #ifdef GPU_DEBUG
-    {
-      __shared__ int n0;
-      if (threadIdx.x == 0)
-        n0 = nloops;
-      __syncthreads();
-      auto ok = n0 == nloops;
-      assert(__syncthreads_and(ok));
-      if (thisModuleId % 100 == 1)
-        if (threadIdx.x == 0)
-          printf("# loops %d\n", nloops);
-    }
+      {
+        if (thisModuleId % 100 == 1)
+            printf("# loops %d\n", nloops);
+      }
 #endif
 
-    __shared__ unsigned int foundClusters;
-    foundClusters = 0;
-    __syncthreads();
+      unsigned int foundClusters = 0;
 
-    // find the number of different clusters, identified by a pixels with clus[i] == i;
-    // mark these pixels with a negative id.
-    for (int i = first; i < msize; i += blockDim.x) {
-      if (id[i] == InvId)  // skip invalid pixels
-        continue;
-      if (clusterId[i] == i) {
-        auto old = atomicInc(&foundClusters, 0xffffffff);
-        clusterId[i] = -(old + 1);
+      // find the number of different clusters, identified by a pixels with clus[i] == i;
+      // mark these pixels with a negative id.
+      for (int i = first; i < msize; ++i) {
+        if (id[i] == InvId)  // skip invalid pixels
+          continue;
+        if (clusterId[i] == i) {
+          auto old = cms::cuda::atomicInc(&foundClusters, 0xffffffff);
```

Comment:
Hmm, shouldn't the wraparound happen implicitly as well, because `foundClusters` is `unsigned int`?

Reply:
Yes, overflow on `unsigned int` will wrap around so this as the same effect as just incrementing by one